classes/class-wc-connect-service-settings-store.php

```diff
@@ -491,6 +491,20 @@ public function get_packages() {
 			return WC_Connect_Options::get_option( 'packages', array() );
 		}
 
+		/**
+		 * Extends the global list of packages with a list of new packages
+		 *
+		 * @param array new_packages - packages to extend
+		 */
+		public function create_packages( $new_packages ) {
+			if (is_null($new_packages)) {
+				return;
+			}
+			$packages = $this->get_packages();
+			$packages = array_merge($packages, $new_packages);
```

Comment:
Thanks for raising this potential duplication issue! I think it'd be better to throw an error in case of duplication, since it's a client side error. I added two commits https://github.com/Automattic/woocommerce-services/pull/2318/commits/901e5d7727acd8a772c14b07d31b59cd01e9ddf5 https://github.com/Automattic/woocommerce-services/pull/2318/commits/ce425b8ba72b7f3ee7d42d1f0ce6fc4a68166065 to handle the following two duplication scenarios:

- The custom/predefined packages in the request body do not have unique names
- Any of the custom/predefined packages have the same names of the existing packages

Sorry that this duplication validation logic gets a bit complicated, I tried adding a test case for each scenario and package type. If you think it's better to handle these duplication scenarios differently, please let me know and I can update the PR accordingly!

Reply:
`$this->settings_store->get_predefined_packages();` and `$this->settings_store->get_packages();` returns an array without a `key`. Finding duplicates will involve looping through the list to map out the `key` we want. I think I would do more or less the same thing you did in the last 2 commits. This looks 👍  to me!